# Analiza danych w języku Python - ćwiczenia laboratoryjne 2022/2023

Ten notatnik zalicza się do grupy zestawów zadań, na podstawie których odbywa się zaliczenie ćwiczeń i podlega zwrotowi do oceny w ustalonym na zajęciach terminie.

Uwagi i wytyczne ogólne dotyczące uzupełniania i oceny notatnika:
- Podczas wykonywania zadań należy korzystać wyłącznie z pakietów zaimportowanych na początku notatnika oraz z pakietów wchodzących w skład standardowej biblioteki Pythona, które można zaimportować samodzielnie we wskazanej komórce notatnika.
- Swoje rozwiązania należy wprowadzać wyłącznie w miejce następujących fragmentów kodu:<br/> `# YOUR CODE HERE`<br/> `raise NotImplementedError()`<br/> 
a odpowiedzi tekstowe w komórkach oznaczonych hasłem:<br/> 
`YOUR ANSWER HERE`<br/> 
Nie należy w żaden sposób modyfikować pozostałych fragmentów kodu oraz innych elementów notatnika, w szczególności dodawać lub usuwać komórek oraz zmieniać nazwy pliku.
- Jeżeli zestaw zadań wymaga skorzystania z fragmentów kodu opracowanego w ramach wcześniejszych zestawów zadań należy je umieścić we wskazanej komórce notatnika.
- Otrzymywane wyniki i odpowiedzi mają być rezultatem wykonania napisanego kodu, odpowiedzi uzupełniane manualnie nie podlegają ocenie.
- Zadanie należy wykonać w taki sposób, aby podczas wykonywania kodu nie zostały wyświetlone żadne ostrzeżenia.
- Zawarte w notatniku automatyczne testy mają charakter poglądowy. Dotyczą one wybranych aspektów zadań i mają za zadanie wyłapać podstawowe błędy. Przejście wszystkich testów nie oznacza, że zadanie jest wykonane w całości poprawnie i zostanie ocenione na maksymalną liczbę punktów.
- Zadania, które powodują wyświetlenie komunikatu o błędzie przerywającym wykonywanie kodu nie podlegają ocenie.

Uwagi i wytyczne ogólne dotyczące wizualizacji wyników:
- Wszystkie wykresy powinny być wykonane w jednolitym, przejrzystym i czytelnym stylu, posiadać odpowiednio dobrane proporcje i zakresy wartości osi.
- Wykresy oraz ich osie powinny mieć nadane tytuły. Jeżeli w obrębie figury znajduje się więcej niż jeden wykres to figura również powinna mieć nadany tytuł. 
- Figury powinny mieć ustawione białe tło, tak, aby niezależnie od ustawień notatnika wszystkie elementy wykresów były dobrze widoczne (domyślnie tło jest przeźroczyste co może powodować problemy w notatnikach z ustawionym ciemnym tłem).
- Rozmiar poziomy figur nie powinien przekraczać 20 cali.

Przed odesłaniem zestawu zadań do oceny proszę uzupełnić komórkę z danymi autorów rozwiązania (`NAME` - nazwa zespołu, `COLLABORATORS` - imiona, nazwiska i numery indeksów członków zespołu) oraz upewnić się, że notatnik działa zgodnie z oczekiwaniami. W tym celu należy skorzystać z opcji **Restart Kernel and Run All Cells...** dostępnej na górnej belce notatnika pod symbolem $\blacktriangleright\blacktriangleright$.

In [1]:
NAME = "IAD21"
COLLABORATORS = "Ewa Szewczyk 406923, Adam Piwowarski 408133"

---

## Zestaw zadań 1: Elementy statystyki opisowej 1 (Notatnik 4/4)

In [2]:
import numpy as np
import pandas as pd
import scipy as sp
import matplotlib as mpl
import matplotlib.pyplot as plt
import seaborn as sns
import dataframe_image as dfi

In [3]:
# Miejsce do importu pakietów wchodzących w skład standardowej biblioteki Pythona oraz ustawienie opcji wykorzystywanych pakietów

### Zadanie 7: Analiza wyników [13 pkt]
Scharakteryzuj wulkanizm ziemski na podstawie wyników analizy przeprowadzonej we wcześniejszych zadaniach zgodnie z poniższymi wytycznymi:
- Umieść treść analizy w pojedynczej komórce Markdown. Oprócz tekstu wklej do niej wszystkie wykresy i tablice, do których odnosi się analiza.
- W analizie oprzyj się przede wszystkim o tablice i grafiki przygotowane w ramach zadań 4-6 (niezbędne minimum).
- Nie ograniczaj się wyłącznie do stwierdzeń typu "zostało zarejestrowanych więcej erupcji o wartości VEI 6 niż 7", tam gdzie to możliwe spróbuj wyciągnąć szersze, dalej idące wnioski o badanym zjawisku.
- Wskaż problemy w analizie przeprowadzonej w ramach wykonanych wcześniej zadań i zaproponuj jakie zmiany należy do niej wprowadzić w celu ich rozwiązania.
- Jeżeli widzisz taką potrzebę możesz wygenerować dodatkowe tablice i grafiki, w takim przypadku generujący je kod powinien zostać umieszczony w komórce poprzedzającej komórkę z treścią analizy.

Wskazówki dotyczące pozyskiwania i wklejania grafik do komórki markdown:
- przygotowane wykresy można zapisać do plików za pomocą kodu: `fig.savefig('image.png')`, gdzie `fig` to zmienna zawierająca wykres przygotowany z wykorzystaniem plt.subplots(), a `'image.png'` to nazwa pliku, do którego ma zostać zapisana.
- przygotowane tablice można zapisać do plików za pomocą kodu: `dfi.export(styled_df, 'df.png')`, gdzie `styled_df` to zmienna zawierająca odpowiednio sformatowaną tablicę, a `'df.png'` to nazwa pliku, do którego ma zostać zapisana.
- zapisane do plików grafiki można wklejać do komórki markdown za pomocą następującej formuły: `![](table.png)`, gdzie `table.png` to nazwa pliku, do którego ma zostać wstawiony do komórki.

Po wygenerowaniu plików z grafikami nie należy uzuwać z notatników kodu wykorzystanego podczas ich przygotowywania, w tym formuł `fig.savefig()` i `dfi.export()`.

In [6]:
# Miejsce na dodatkowy kod
# YOUR CODE HERE

eruptions_data = pd.read_csv('eruptions_data.csv', index_col='Eruption Number')
most_active_data = eruptions_data['Volcano Name'].value_counts().reset_index()
most_active_data = most_active_data.rename(columns={'index':'Volcano', 'Volcano Name': 'Recorded eruptions'}) \
    .sort_values(by=['Recorded eruptions', 'Volcano'], ascending=[False, True])

print(most_active_data['Recorded eruptions'].mean())

observations = eruptions_data[eruptions_data['Evidence Type'] == 'Observations']['Volcano Name'].count()
print(observations)

# raise NotImplementedError()

11.603641456582633
6187


YOUR ANSWER HERE
# Analiza aktywności sejsmicznej
### Ogólniki i ciekawostki
![](10_most_active.png)
<br />*Tabela zawierająca 10 najaktywniejszych wulkanów*

<br />Zgodnie z powyższą tabelą wulkany o najwyższej aktywności wybuchały nawet 191 razy (Fournaise, Piton de la), co jest wynikiem znacząco wyższym od średniej wynoszącej 11.6 razy.<br />

![](last_eruption_data.png)
<br />*Lata najnowszych erupcji pogrupowane wg skali VEI*

<br />Siła erupcji określana za pomocą VEI w skali od 0 do 7 jest prawdopodobnie najistotniejszym elementem analizy. Jak przedstawia powyższa tabela najsilniejsze erupcje wystąpiły w naszym zbiorze danych ponad 30, a najsilniejsza (VEI=7) ponad 200 lat temu. <br /> 

### Analiza
W tej sekcji przedstawiona zostanie analiza częstości występowania erupcji oraz metod ich wykrywania uwzględniając podział według skali VEI.

![](vei_data.png)
<br /> *Tabela zawierająca odpowiednio zliczenia, zliczenia skumulowane, częstości i częstości skumulowane erupcji z podziałem na ich kategorie oraz siłę*
![](02_szeregi.png)
<br /> *Wykresy ilości erupcji z podziałem wg skali VEI z nałożoną linią częstości skumulowanej; od lewej: wszystkie erupcje, potwierdzone, niepotwierdzone*

Przechodząc do analizy częstości występowania erupcji widzimy, że najczętszymi erupcjami są te o VEI równym 2 - aż 3991 takich przypadków. Na wszystkich trzech wykresach można zauważyć, że częstość skumulowana zliczeń rośnie gwałtownie, aż do VEI równego 3 - odtąd występuje wypłaszczenie. Powinna to być dla nas dobra informacja: silniejsze erupcje (o VEI>3) zdarzają się zauważalnie rzadziej.<br /> 
Dodatkowo zgodnie z tabelą oraz pierwszym od prawej wykresem, prawie 8% wszystkich erupcji jest niepotwierdzone. W związku z tym możemy nie uwzględniać tych danych w ewentualnych dalszych analizach - mogą to być dane przypadkowe/błędne powodujące zaburzenia różnych badanych cech i właściwości. Nie ma za to niepotwierdzonych erupcji o VEI > 4. Wynika to prawdopodobnie z tego, że silniejsza erupcja powinna być zarejestrowana przez więcej czujników czy też po prostu bardziej zauważona przez ludzi, a jeśli chodzi o te z przeszłości - zapewne pozostawiłaby odpowiednią ilość śladów.<br />

---
 <br /> 

![](evidence_data.png)
<br /> *Tabela zestawiająca ze sobą wszystkie typy i metody ewidencjonowania erupcji z podziałem wg skali VEI*
![](03_metody.png)
<br />*Wizualizacja rozkładu zarejestrowanych erupcji w czasie (dolna dodatkowo z podziałem na wykorzystane metody*

Biorąc pod uwagę typy ewidencjonowania, zdecydowanie najpopularniejszym jest zwykła obserwacja - aż 6187 przypadków. Natomiast w wykrywaniu erupcji z przeszłości prym wiodą metody izotopowe, a wśród nich metoda 14C (skalibrowana). Jak możemy zauważyć na powyższym wykresie, w ostatnich latach główną metodą ewidencjonowania jest zwykła obserwacja. Wynika to najprawdopodobniej z tego, że w dzisiejszych czasach wiele erupcji jest zauważanych przez wyspecjalizowane do tego czujniki rozlokawane na odpowiednich obszarach, które natychmiast raportują nawet najmniejsze drgania. Kiedyś ludzkość nie dysponowała takimi metodami oraz po prostu była mniej świadoma - zapewne niektóre erupcje mogły odbyć się "niepostrzeżenie" lub być odczytana jako np. gniew bogów. Nawet jeśli były one jakoś ewidencjonowane, niewiele z tych dokumentów przetrwało do dziś.

### Co można poprawić
Aby zwiększyć ciekawość tej analizy można by dodać dokładniejsze zbadanie metod ewidencjonowania w zależności od początkowego roku erupcji - przyniosłoby to istotne dla niezaznajomionych z sejsmiką osób informacje, kiedy jaką metodę się wykorzystuje. Dla starszych erupcji, jeśli pozwoliłyby dane, można by również badać użytą metodę w zależności od roku startowego i końcowego - czasu trwania drgań.<br />
Ciekawe byłoby również sprawdzenie, które wulkany najczęściej wybuchają oraz, czy każdy wulkan ma jakąś swoją charakterystykę, np. erupcje są słabe, ale bardzo rozległe w czasie, czy też szybkie i gwałtowne - czy może jest to kwestia losowa.